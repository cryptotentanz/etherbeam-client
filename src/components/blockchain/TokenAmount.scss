@use '../../styles/resources/fonts';

.my-tokenamount {
  &-symbol {
    font-family: fonts.$primary;
    font-weight: bold;
  }

  &-amount {
    font-family: fonts.$monospace;
  }
}
