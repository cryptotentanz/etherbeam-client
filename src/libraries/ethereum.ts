import { Address, AddressType } from './types'

const ETHERSCAN_URL_BASE = 'https://etherscan.io/'

export type EtherscanUrlType = 'block' | 'tx' | 'token' | 'address'

export const getEtherscanUrl = (type: EtherscanUrlType, id: string | number): string => {
  return `${ETHERSCAN_URL_BASE}${type}/${id}`
}

export const createTransactionAddress = (hash: string): Address => {
  return {
    hash,
    type: AddressType.Transaction,
    url: getEtherscanUrl('tx', hash),
  }
}
