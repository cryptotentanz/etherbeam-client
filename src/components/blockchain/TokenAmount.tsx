import React, { FC, ReactNode } from 'react'
import { Token } from 'libraries/types'
import { BigNumber } from 'ethers'
import { AmountPrintOptions, printEtherAmount, printTokenAmount } from './helpers'
import Badge from '../Badge'
import './TokenAmount.scss'

const ETH_SYMBOL = 'ETH'

type ColorProp = 'dark' | 'light' | 'primary' | 'secondary' | 'positive' | 'negative'
type SizeProp = 's' | 'm' | 'l' | 'xl'

interface TokenAmountProps {
  amount: BigNumber | null | undefined
  token?: Token
  color?: ColorProp
  size?: SizeProp
  amountPrintOptions?: AmountPrintOptions
  noSymbol?: boolean
  noBadge?: boolean
  className?: string
}

const TokenAmount: FC<TokenAmountProps> = ({
  amount,
  token,
  color = 'light',
  size,
  amountPrintOptions,
  noSymbol,
  noBadge,
  className,
}: TokenAmountProps) => {
  const getPrintedAmount = () => {
    if (!amount) return '--'

    return token ? printTokenAmount(amount, token, amountPrintOptions) : printEtherAmount(amount, amountPrintOptions)
  }

  const renderAmount = (content: ReactNode) => {
    if (noBadge) return <span className={className}>{content}</span>

    return (
      <Badge color={color} size={size} className={className}>
        {content}
      </Badge>
    )
  }

  return (
    <>
      {renderAmount(
        <>
          {!noSymbol && <span className="my-tokenamount-symbol">{token ? token.symbol : ETH_SYMBOL}</span>}
          <span className="my-tokenamount-amount"> {getPrintedAmount()}</span>
        </>
      )}
    </>
  )
}

export default TokenAmount
