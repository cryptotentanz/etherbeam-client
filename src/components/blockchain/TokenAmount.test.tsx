import * as React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import { createToken } from 'tests/fixtures/ethereum'
import { BigNumber } from 'ethers'
import TokenAmount from './TokenAmount'

describe('<TokenAmount />', () => {
  const className = 'my-class'
  let component: ShallowWrapper

  describe('Ether', () => {
    beforeAll(() => {
      component = shallow(<TokenAmount amount={BigNumber.from('1000000000000000000000')} />)
    })

    it('should render', () => {
      const badge = component.find('Badge')
      expect(badge.prop('color')).toBe('light')

      const spans = badge.find('span')
      expect(spans.at(0).hasClass('my-tokenamount-symbol')).toBeTruthy()
      expect(spans.at(0).prop('children')).toEqual('ETH')
      expect(spans.at(1).hasClass('my-tokenamount-amount')).toBeTruthy()
      expect((spans.at(1).prop('children') as never)?.[1]).toEqual('1,000')
    })

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Token', () => {
    const token = createToken('0x0000000000000000000000000000000000000111')

    beforeAll(() => {
      component = shallow(<TokenAmount amount={BigNumber.from('1000000000000000000000')} token={token} />)
    })

    it('should render', () => {
      const spans = component.find('Badge').find('span')
      expect(spans.at(0).prop('children')).toEqual(token.symbol)
      expect((spans.at(1).prop('children') as never)?.[1]).toEqual('1,000')
    })
  })

  describe('Color', () => {
    beforeAll(() => {
      component = shallow(<TokenAmount amount={BigNumber.from('1000000000000000000000')} color="primary" />)
    })

    it('should render', () => expect(component.find('Badge').prop('color')).toBe('primary'))
  })

  describe('Size', () => {
    beforeAll(() => {
      component = shallow(<TokenAmount amount={BigNumber.from('1000000000000000000000')} size="l" />)
    })

    it('should render', () => expect(component.find('Badge').prop('size')).toBe('l'))
  })

  describe('Print options', () => {
    beforeAll(() => {
      component = shallow(
        <TokenAmount
          amount={BigNumber.from('1000000000000000000000')}
          amountPrintOptions={{ decimals: 2, forceDecimals: true }}
        />
      )
    })

    it('should render', () => {
      const spans = component.find('Badge').find('span')
      expect((spans.at(1).prop('children') as never)?.[1]).toEqual('1,000.00')
    })
  })

  describe('No symbol', () => {
    beforeAll(() => {
      component = shallow(<TokenAmount amount={BigNumber.from('1000000000000000000000')} noSymbol />)
    })

    it('should render', () => {
      const spans = component.find('Badge').find('span')
      expect(spans.at(0).hasClass('my-tokenamount-amount')).toBeTruthy()
      expect((spans.at(0).prop('children') as never)?.[1]).toEqual('1,000')
    })
  })

  describe('Class', () => {
    beforeAll(() => {
      component = shallow(<TokenAmount amount={BigNumber.from('1000000000000000000000')} className={className} />)
    })

    it('should render', () => expect(component.find('Badge').prop('className')).toBe(className))
  })
})
