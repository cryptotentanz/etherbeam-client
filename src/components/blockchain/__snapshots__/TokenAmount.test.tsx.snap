// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TokenAmount /> Ether should match snapshot 1`] = `
<span
  class="my-badge my-light-bg  "
>
  <span
    class="my-tokenamount-symbol"
  >
    ETH
  </span>
  <span
    class="my-tokenamount-amount"
  >
     1,000
  </span>
</span>
`;
